framework/web/UrlNormalizer.php

```diff
@@ -0,0 +1,151 @@
+<?php
+
+/**
+ * @link http://www.yiiframework.com/
+ * @copyright Copyright (c) 2008 Yii Software LLC
+ * @license http://www.yiiframework.com/license/
+ */
+
+namespace yii\web;
+
+use Yii;
+use yii\base\Object;
+use yii\base\InvalidConfigException;
+
+/**
+ * UrlNormalizer normalizes URLs for [[UrlManager]] and [[UrlRule]].
+ *
+ * @author Robert Korulczyk <robert@korulczyk.pl>
+ * @author Cronfy <cronfy@gmail.com>
+ * @since 2.0.10
+ */
+class UrlNormalizer extends Object
+{
+    /**
+     * Represents permament redirection during route normalization.
+     * @see https://en.wikipedia.org/wiki/HTTP_301
+     */
+    const ACTION_REDIRECT_PERMANENT = 301;
+    /**
+     * Represents temporary redirection during route normalization.
+     * @see https://en.wikipedia.org/wiki/HTTP_302
+     */
+    const ACTION_REDIRECT_TEMPORARY = 302;
+    /**
+     * Represents showing 404 error page during route normalization.
+     * @see https://en.wikipedia.org/wiki/HTTP_404
+     */
+    const ACTION_NOT_FOUND = 404;
+
+    /**
+     * @var boolean whether slashes should be collapsed, for example `site///index` will be
+     * converted into `site/index`
+     */
+    public $collapseSlashes = true;
+    /**
+     * @var boolean whether trailing slash should be normalized according to the suffix settings
+     * of the rule
+     */
+    public $normalizeTrailingSlash = true;
+    /**
+     * @var integer|callable|null action to perform during route normalization.
+     * Available options are:
+     * - `null` - no special action will be performed
+     * - `301` - the request should be redirected to the normalized URL using
+     * permanent redirection
+     * - `302` - the request should be redirected to the normalized URL using
+     * temporary redirection
+     * - `404` - [[NotFoundHttpException]] will be thrown
+     * - `callable` - custom user callback, for example:
+     *
+     *   ```php
+     *   function ($route, $normalizer) {
+     *       // use custom action for redirections
+     *       $route[1]['oldRoute'] = $route[0];
+     *       $route[0] = 'site/redirect';
+     *       return $route;
+     *   }
+     *   ```
+     */
+    public $action = self::ACTION_REDIRECT_PERMANENT;
```

Comment:
permanent redirection can be a real pain when developing as they are cached in the browser. Maybe we should make this value depend on YII_DEBUG?